[BUG] "Execution context was destroyed" when using Locator.isVisible()

<!-- ⚠️⚠️ Do not delete this template ⚠️⚠️ -->

<!-- 🔎 Search existing issues to avoid creating duplicates. -->
<!-- 🧪 Test using the latest Playwright release to see if your issue has already been fixed -->
<!-- 💡 Provide enough information for us to be able to reproduce your issue locally -->

### System info
- Playwright Version: [v1.32.0 (Java)]
- Operating System: [macOS 13.3]
- Browser: [Firefox, WebKit, (Chromium unclear)]
- Other info:

### Source code

- [ ] I provided exact source code that allows reproducing the issue locally.

<!-- For simple cases, please provide a self-contained test file along with the config file -->
<!-- For larger cases, you can provide a GitHub repo you created for this issue -->
<!-- If we can not reproduce the problem locally, we won't be able to act on it -->
<!-- You can still file without the exact code and we will try to help, but if we can't repro, it will be closed -->


My website asynchronously navigates to another page while I am trying to use locators like this in a loop (because I am waiting for one of two locators to become visible):
```java
navbar
    .locator(LOGGED_IN_NAVBAR_SELECTOR)
    .locator(ACCOUNT_MENU_BUTTON_SELECTOR)
    .isVisible();
```

This results in an error "Execution context was destroyed". I have reproduced this on Webkit and Firefox, but could not manage to reproduce it with Chromium. It's not happening every time because of the racy nature of the async navigation. I don't know if Chromium doesn't have this problem or the timing is just different, thereby avoiding the error.

Stack trace on Webkit:
```
com.microsoft.playwright.PlaywrightException: Error {
  message='Execution context was destroyed, most likely because of a navigation
=========================== logs ===========================
  checking visibility of locator("app-navbar").locator("app-logged-in-navbar").locator("button:has-text('arrow_drop_down')")
============================================================
  name='Error
  stack='Error: Execution context was destroyed, most likely because of a navigation
=========================== logs ===========================
  checking visibility of locator("app-navbar").locator("app-logged-in-navbar").locator("button:has-text('arrow_drop_down')")
============================================================
    at Frame._contextDestroyed (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/frames.js:1446:30)
    at WKPage._removeContextsForFrame (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/webkit/wkPage.js:453:32)
    at WKPage._onFrameNavigated (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/webkit/wkPage.js:438:10)
    at WKSession.<anonymous> (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/webkit/wkPage.js:389:127)
    at WKSession.emit (node:events:513:28)
    at /private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/webkit/wkConnection.js:172:41
    at runNextTicks (node:internal/process/task_queues:60:5)
    at process.processImmediate (node:internal/timers:447:9)
}

	at com.microsoft.playwright.impl.WaitableResult.get(WaitableResult.java:54)
	at com.microsoft.playwright.impl.ChannelOwner.runUntil(ChannelOwner.java:111)
	at com.microsoft.playwright.impl.Connection.sendMessage(Connection.java:126)
	at com.microsoft.playwright.impl.ChannelOwner.sendMessage(ChannelOwner.java:102)
	at com.microsoft.playwright.impl.FrameImpl.isVisibleImpl(FrameImpl.java:641)
	at com.microsoft.playwright.impl.FrameImpl.lambda$isVisible$29(FrameImpl.java:627)
	at com.microsoft.playwright.impl.LoggingSupport.withLogging(LoggingSupport.java:47)
	at com.microsoft.playwright.impl.ChannelOwner.withLogging(ChannelOwner.java:87)
	at com.microsoft.playwright.impl.FrameImpl.isVisible(FrameImpl.java:627)
	at com.microsoft.playwright.impl.LocatorImpl.isVisible(LocatorImpl.java:374)
	at com.microsoft.playwright.Locator.isVisible(Locator.java:3668)
        <... project specific frames ...>
Caused by: com.microsoft.playwright.impl.DriverException: Error {
  message='Execution context was destroyed, most likely because of a navigation
=========================== logs ===========================
  checking visibility of locator("app-navbar").locator("app-logged-in-navbar").locator("button:has-text('arrow_drop_down')")
============================================================
  name='Error
  stack='Error: Execution context was destroyed, most likely because of a navigation
=========================== logs ===========================
  checking visibility of locator("app-navbar").locator("app-logged-in-navbar").locator("button:has-text('arrow_drop_down')")
============================================================
    at Frame._contextDestroyed (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/frames.js:1446:30)
    at WKPage._removeContextsForFrame (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/webkit/wkPage.js:453:32)
    at WKPage._onFrameNavigated (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/webkit/wkPage.js:438:10)
    at WKSession.<anonymous> (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/webkit/wkPage.js:389:127)
    at WKSession.emit (node:events:513:28)
    at /private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-4016531511947763690/package/lib/server/webkit/wkConnection.js:172:41
    at runNextTicks (node:internal/process/task_queues:60:5)
    at process.processImmediate (node:internal/timers:447:9)
}
	at com.microsoft.playwright.impl.Connection.dispatch(Connection.java:226)
	at com.microsoft.playwright.impl.Connection.processOneMessage(Connection.java:206)
	at com.microsoft.playwright.impl.ChannelOwner.runUntil(ChannelOwner.java:109)
	... 53 more
```


Firefox stack trace:
```
com.microsoft.playwright.PlaywrightException: Error {
  message='Execution context was destroyed, most likely because of a navigation
=========================== logs ===========================
  checking visibility of locator("app-navbar").locator("app-logged-in-navbar").locator("button:has-text('arrow_drop_down')")
============================================================
  name='Error
  stack='Error: Execution context was destroyed, most likely because of a navigation
=========================== logs ===========================
  checking visibility of locator("app-navbar").locator("app-logged-in-navbar").locator("button:has-text('arrow_drop_down')")
============================================================
    at Frame._contextDestroyed (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-12798600472171551690/package/lib/server/frames.js:1446:30)
    at FFPage._onExecutionContextDestroyed (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-12798600472171551690/package/lib/server/firefox/ffPage.js:141:19)
    at FFPage._onExecutionContextsCleared (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-12798600472171551690/package/lib/server/firefox/ffPage.js:144:88)
    at FFSession.emit (node:events:513:28)
    at /private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-12798600472171551690/package/lib/server/firefox/ffConnection.js:204:41
    at runNextTicks (node:internal/process/task_queues:60:5)
    at process.processImmediate (node:internal/timers:447:9)
}

	at com.microsoft.playwright.impl.WaitableResult.get(WaitableResult.java:54)
	at com.microsoft.playwright.impl.ChannelOwner.runUntil(ChannelOwner.java:111)
	at com.microsoft.playwright.impl.Connection.sendMessage(Connection.java:126)
	at com.microsoft.playwright.impl.ChannelOwner.sendMessage(ChannelOwner.java:102)
	at com.microsoft.playwright.impl.FrameImpl.isVisibleImpl(FrameImpl.java:641)
	at com.microsoft.playwright.impl.FrameImpl.lambda$isVisible$29(FrameImpl.java:627)
	at com.microsoft.playwright.impl.LoggingSupport.withLogging(LoggingSupport.java:47)
	at com.microsoft.playwright.impl.ChannelOwner.withLogging(ChannelOwner.java:87)
	at com.microsoft.playwright.impl.FrameImpl.isVisible(FrameImpl.java:627)
	at com.microsoft.playwright.impl.LocatorImpl.isVisible(LocatorImpl.java:374)
	at com.microsoft.playwright.Locator.isVisible(Locator.java:3668)
	< ... project specific frames ... >
Caused by: com.microsoft.playwright.impl.DriverException: Error {
  message='Execution context was destroyed, most likely because of a navigation
=========================== logs ===========================
  checking visibility of locator("app-navbar").locator("app-logged-in-navbar").locator("button:has-text('arrow_drop_down')")
============================================================
  name='Error
  stack='Error: Execution context was destroyed, most likely because of a navigation
=========================== logs ===========================
  checking visibility of locator("app-navbar").locator("app-logged-in-navbar").locator("button:has-text('arrow_drop_down')")
============================================================
    at Frame._contextDestroyed (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-12798600472171551690/package/lib/server/frames.js:1446:30)
    at FFPage._onExecutionContextDestroyed (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-12798600472171551690/package/lib/server/firefox/ffPage.js:141:19)
    at FFPage._onExecutionContextsCleared (/private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-12798600472171551690/package/lib/server/firefox/ffPage.js:144:88)
    at FFSession.emit (node:events:513:28)
    at /private/var/folders/l5/mt5xnj_14rn9j4lsnfs5n9b40000gn/T/playwright-java-12798600472171551690/package/lib/server/firefox/ffConnection.js:204:41
    at runNextTicks (node:internal/process/task_queues:60:5)
    at process.processImmediate (node:internal/timers:447:9)
}
	at com.microsoft.playwright.impl.Connection.dispatch(Connection.java:226)
	at com.microsoft.playwright.impl.Connection.processOneMessage(Connection.java:206)
	at com.microsoft.playwright.impl.ChannelOwner.runUntil(ChannelOwner.java:109)
	... 53 more
```

This is actually the same issue as https://github.com/microsoft/playwright/issues/16720
The solution there was to use `assertThat(locator).isVisible`.

This doesn't work here, because I am manually waiting for one of two conditions to occur, so I need to use `Locator.isVisible`.

Previously this was "solved" (?) by putting a `Page.waitForLoadState(LoadState.NETWORKIDLE)` in front of the visibility checking loop. This presumably worked because the async navigation happens shortly after loading the page, so it would have happened during the NETWORKIDLE waiting, after which the execution context is stable.

I now removed the `waitForLoadState` because NETWORKIDLE is discouraged, and asserting for web elements is recommended, per https://github.com/microsoft/playwright/issues/22897#issuecomment-1540452266

But then this bug got triggered instead.

--- a/packages/playwright-core/src/server/frames.ts
+++ b/packages/playwright-core/src/server/frames.ts
@@ -1288,7 +1288,11 @@ export class Frame extends SdkObject {
         const state = element ? injected.elementState(element, 'visible') : false;
         return state === 'error:notconnected' ? false : state;
       }, { info: resolved.info });
-    }, this._page._timeoutSettings.timeout({}));
+    }, this._page._timeoutSettings.timeout({})).catch(e => {
+      if (js.isJavaScriptErrorInEvaluate(e) || isInvalidSelectorError(e) || isSessionClosedError(e))
+        throw e;
+      return false;
+    });
   }
 
   async isHidden(metadata: CallMetadata, selector: string, options: types.StrictOptions = {}): Promise<boolean> {

--- a/packages/playwright-core/src/utils/manualPromise.ts
+++ b/packages/playwright-core/src/utils/manualPromise.ts
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+import { rewriteErrorMessage } from './stackTrace';
+
 export class ManualPromise<T = void> extends Promise<T> {
   private _resolve!: (t: T) => void;
   private _reject!: (e: Error) => void;
@@ -56,12 +58,14 @@ export class ManualPromise<T = void> extends Promise<T> {
 
 export class ScopedRace {
   private _terminateError: Error | undefined;
-  private _terminatePromises = new Set<ManualPromise<Error>>();
+  private _terminatePromises = new Map<ManualPromise<Error>, Error>();
 
   scopeClosed(error: Error) {
     this._terminateError = error;
-    for (const p of this._terminatePromises)
-      p.resolve(error);
+    for (const [p, e] of this._terminatePromises) {
+      rewriteErrorMessage(e, error.message);
+      p.resolve(e);
+    }
   }
 
   async race<T>(promise: Promise<T>): Promise<T> {
@@ -76,7 +80,8 @@ export class ScopedRace {
     const terminatePromise = new ManualPromise<Error>();
     if (this._terminateError)
       terminatePromise.resolve(this._terminateError);
-    this._terminatePromises.add(terminatePromise);
+    const error = new Error('');
+    this._terminatePromises.set(terminatePromise, error);
     try {
       return await Promise.race([
         terminatePromise.then(e => safe ? defaultValue : Promise.reject(e)),

--- a/tests/page/locator-is-visible.spec.ts
+++ b/tests/page/locator-is-visible.spec.ts
@@ -85,3 +85,23 @@ it('isVisible inside a role=button', async ({ page }) => {
   await span.waitFor({ state: 'hidden' });
   await page.locator('[role=button]').waitFor({ state: 'visible' });
 });
+
+it('isVisible during navigation should not throw', async ({ page, server }) => {
+  for (let i = 0; i < 20; i++) {
+    // Make sure previous navigation finishes, to avoid page.setContent throwing.
+    await page.waitForTimeout(100);
+    await page.setContent(`
+      <script>
+        setTimeout(() => {
+          window.location.href = ${JSON.stringify(server.EMPTY_PAGE)};
+        }, Math.random(50));
+      </script>
+    `);
+    expect(await page.locator('div').isVisible()).toBe(false);
+  }
+});
+
+it('isVisible with invalid selector should throw', async ({ page }) => {
+  const error = await page.locator('hey=what').isVisible().catch(e => e);
+  expect(error.message).toContain('Unknown engine "hey" while parsing selector hey=what');
+});

